rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,356 @@
+expired. An error will be returned if the upload has already been completed. The user must own the collection and be
+the one who initiated the upload in order to retrieve an additional access token.
+
+If the size of the file being upload is <5GB, then a presigned URL will be returned.
+If the size of the file is >5GB, then a access token will be returned.
+
+##### Request: POST collection/{collection_id}/{dataset_id}/upload/file
+
+- File size - the size of the file to be uploaded
+- collection_id - identifies the collection
+- attestation - A boolean value indicating if the file is for attestation
+- **Optional** asset_uuid - identifies the name of the new file created.
+
+##### Response: POST collection/{collection_id}/{dataset_id}/upload/file
+
+- Access Token - allows the user to upload their file to S3 in multiple parts.
+- Presigned URL - allows the user to upload their file to S3 in a single part.
+- asset_uuid - identifies the name of the new file created.
+
+TODO: add error respones
+
+### Upload Flows
+
+There are two methods for uploading data to the DP. They are Cloud to Cloud and Local to Cloud. The follow
+sections details the user interactions required, and how the Data Portal backend will facilitate the upload.
+
+#### Cloud to Cloud Upload Flow
+
+Cloud to cloud uploads require the user to provide a shareable link for a file on their CSP from which DP can upload
+the file to the S3 bucket. Now to walk though the cloud to cloud flow, see figure 1.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 2:** Architecture of a user uploading from their cloud to our cloud.
+
+##### 1. Get Share Link
+
+The user must generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
+sharable links. Only sharable links should only point to a single file, not a folder or collection of files.
+
+##### 2. Share Link
+
+Using the browser App, make a _POST {collection_id}/{dataset_id}/upload/link_ request to upload the file
+by providing the sharable link in the request body. The response will contain the asset uuid, that can be used to
+check the status of the upload
+
+##### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
+links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of sanitization. The backend that runs this should also be given minimal privileges to further protect from
+server side request forgery.
+
+###### How to download from a shared link
+
+- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
+- Google drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>
+
+##### 4. _Optional_ Queue
```

Comment:
Curious -- why is this optional? If we don't build this, how do we handle errors?

Reply:
We could put all of the retry logic into the EC2 instances.

Reply:
I'd advocate for this queue -- seems like a cleaner separation of concerns plus you get the deadletter queue for free.

Reply:
Also curious, why EC2 vs something like SQS + Lambda? Might make things easier to manage, separate the download contexts, and give you more resilience/scale. If payloads are capped at 30GB this might be workable within the lambda bounds.